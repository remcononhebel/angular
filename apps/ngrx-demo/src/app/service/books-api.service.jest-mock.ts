import { Provider } from '@angular/core';
import { of } from 'rxjs';
import { mockBooks } from '../mocks/book.mocks';
import { BooksApiService } from './books-api.service';

export class BooksApiServiceJestMock {
  getBooks = jest.fn().mockReturnValue(of({ items: [...mockBooks] }));

  static provide(): Provider {
    return { provide: BooksApiService, useClass: BooksApiServiceJestMock };
  }
}
