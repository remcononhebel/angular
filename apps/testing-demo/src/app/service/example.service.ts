import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';

export const availableFruits = [
  'Apple',
  'Banana',
  'Orange',
  'Grapes',
  'Mango',
  'Strawberry',
];


@Injectable({
  providedIn: 'root',
})
export class ExampleService {

  readonly selectedFruit$ = new Subject<string>();

  getValue(): number {
    return 42;
  }

  selectRandomFruit(): void {
    const randomIndex = Math.floor(Math.random() * availableFruits.length);
    this.selectedFruit$.next(availableFruits[randomIndex]);
  }
}
