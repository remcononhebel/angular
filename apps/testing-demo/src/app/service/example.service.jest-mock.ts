import { Provider } from '@angular/core';
import { Spectator } from '@ngneat/spectator';
import { ExampleService } from './example.service';
import { Subject } from 'rxjs';

type ExampleServiceInterface = Pick<ExampleService, keyof ExampleService>;

export class ExampleServiceJestMock implements ExampleServiceInterface {
  static readonly fruits = [
    'Apple',
    'Banana',
    'Orange',
    'Grapes',
    'Mango',
    'Strawberry',
  ];

  selectedFruit$ = new Subject<string>();

  getValue = jest.fn();
  selectRandomFruit = jest.fn();

  static inject(spectator: Spectator<unknown>): ExampleServiceJestMock {
    return spectator.inject(
      ExampleService
    ) as ExampleServiceInterface as ExampleServiceJestMock;
  }

  static provide(): Provider {
    return { provide: ExampleService, useClass: ExampleServiceJestMock };
  }
}
