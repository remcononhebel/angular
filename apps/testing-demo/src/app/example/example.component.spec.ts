import { Spectator, createComponentFactory } from '@ngneat/spectator';
import { first } from 'rxjs';
import { ExampleServiceJestMock } from '../service/example.service.jest-mock';
import { ExampleComponent } from './example.component';

describe('ExampleComponent', () => {
  let spectator: Spectator<ExampleComponent>;
  let component: ExampleComponent;

  const createComponent = createComponentFactory({
    component: ExampleComponent,
    providers: [ExampleServiceJestMock.provide()],
  });

  beforeEach(() => {
    spectator = createComponent();
    component = spectator.component;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('getValue', () => {
    it('should not render a result by default', () => {
      expect(spectator.query('[data-testid="result"]')).toBeFalsy();
    });

    it('should call ExampleService.getValue() when button is clicked', () => {
      const exampleService = ExampleServiceJestMock.inject(spectator);
      const spy = jest.spyOn(exampleService, 'getValue');

      spectator.click('[data-testid="button"]');

      expect(spy).toHaveBeenCalled();
    });

    it('should get a value from the ExampleService when getValue is called', (done: jest.DoneCallback) => {
      const mockValue = 17;

      const exampleService = ExampleServiceJestMock.inject(spectator);
      exampleService.getValue.mockImplementation(() => {
        console.warn('ExampleServiceJestMock.getValue called!');
        return mockValue;
      });

      component.serviceValue$.pipe(first()).subscribe((value: number) => {
        expect(value).toEqual(mockValue);
        done();
      });

      component.getValue();
    });
  });

  describe('selectedFruit', () => {
    it('should not render a selected fruit by default', () => {
      expect(spectator.query('[data-testid="selected-fruit"]')).toBeFalsy();
    });

    it('should not render a selected fruit by default (snapshot)', () => {
      expect(spectator.fixture).toMatchSnapshot();
    });

    it('should call the ExampleService.selectRandomFruit when selectRandomFruit is called', () => {
      const exampleService = ExampleServiceJestMock.inject(spectator);
      const spy = jest.spyOn(exampleService, 'selectRandomFruit');

      component.selectRandomFruit();

      expect(spy).toHaveBeenCalled();
    });

    it('should render a fruit when selectedFruit$ is emitted by the ExampleService', (done: jest.DoneCallback) => {
      const exampleService = ExampleServiceJestMock.inject(spectator);

      exampleService.selectedFruit$.pipe(first()).subscribe(() => {
        spectator.detectChanges();
        expect(spectator.query('[data-testid="selected-fruit"]')).toBeTruthy();

        done();
      });

      exampleService.selectedFruit$.next('mock fruit');
    });

    it('should render a fruit when selectedFruit$ is emitted by the ExampleService (snapshot)', (done: jest.DoneCallback) => {
      const exampleService = ExampleServiceJestMock.inject(spectator);

      exampleService.selectedFruit$.pipe(first()).subscribe(() => {
        spectator.detectChanges();
        expect(spectator.fixture).toMatchSnapshot();

        done();
      });

      exampleService.selectedFruit$.next('mock fruit');
    });
  });
});
