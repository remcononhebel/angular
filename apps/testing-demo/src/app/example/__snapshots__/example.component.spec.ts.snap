// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExampleComponent selectedFruit should not render a selected fruit by default (snapshot) 1`] = `
<testing-demo-example
  exampleService={[Function ExampleServiceJestMock]}
  serviceValue$={[Function Subject]}
>
  <h2>
    Example component
  </h2><div>
    <button
      data-testid="button"
    >
      Get value
    </button>
    <p>
      No value available
    </p>
  </div><div>
    <button
      data-testid="fruit-button"
    >
       Select random fruit 
    </button>
    <p>
      No fruit selected
    </p>
  </div>
</testing-demo-example>
`;

exports[`ExampleComponent selectedFruit should render a fruit when selectedFruit$ is emitted by the ExampleService (snapshot) 1`] = `
<testing-demo-example
  exampleService={[Function ExampleServiceJestMock]}
  serviceValue$={[Function Subject]}
>
  <h2>
    Example component
  </h2><div>
    <button
      data-testid="button"
    >
      Get value
    </button>
    <p>
      No value available
    </p>
  </div><div>
    <button
      data-testid="fruit-button"
    >
       Select random fruit 
    </button>
    <p
      data-testid="selected-fruit"
    >
       Selected fruit: mock fruit 
    </p>
  </div>
</testing-demo-example>
`;
